File: server_integration_tests/clients/java_reactive_client/client-app/src/test/java/dev/graphlink/server/InterceptTest.java
package dev.graphlink.server;

import dev.graphlink.server.generated.client.GraphLinkClient;
import dev.graphlink.server.generated.client.GraphLinkException;
import dev.graphlink.server.generated.inputs.ListTeamsFieldArgs;
import dev.graphlink.server.generated.types.GreetingReceivedResponse;
import dev.graphlink.server.generated.types.Member;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Test;

import java.util.Set;
import java.util.concurrent.CompletableFuture;
import java.util.concurrent.TimeUnit;
import java.util.stream.Collectors;

import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.assertThrows;

public class InterceptTest {

    private GraphLinkClient client;

    @BeforeEach
    void setUp() {
        client = Fixtures.newClient();
    }

    @Test
    void greetInterceptedAppendsTheRunBeforeMarkerToTheNameArgument() {
        var res = client.queries.greetIntercepted("Ada").block();
        assertEquals("Hello, Ada(runBefore)", res.getGreetIntercepted().getMessage());
    }

    @Test
    void greetPlainIsNotInterceptedNoMarkerAppended() {
        var res = client.queries.greetPlain("Ada").block();
        assertEquals("Hello, Ada", res.getGreetPlain().getMessage());
    }

    @Test
    void greetingReceivedBakesTheMarkerIntoEveryEmittedGreeting() throws Exception {
        CompletableFuture<GreetingReceivedResponse> eventFuture =
                client.subscriptions.greetingReceived().next().toFuture();

        Thread.sleep(300);

        client.mutations.sendGreeting("Ada").block();

        var event = eventFuture.get(10, TimeUnit.SECONDS);
        assertEquals("Hello, Ada(runBefore)", event.getGreetingReceived().getMessage());
    }

    @Test
    void teamMembersInterceptedMappingAllowsRoleAuth() {
        var res = client.queries.listTeams(new ListTeamsFieldArgs("auth")).block();
        var members = res.getListTeams().get(0).getMembers().stream().map(Member::getName).collect(Collectors.toSet());
        assertEquals(Set.of("Ada", "Grace"), members);
    }

    @Test
    void teamMembersInterceptedMappingDeniesAnyOtherRole() {
        assertThrows(GraphLinkException.class,
                () -> client.queries.listTeams(new ListTeamsFieldArgs("wrong")).block());
    }
}

File: server_integration_tests/clients/swift_client/Tests/GraphLinkGeneratedTests/InterceptTest.swift
import XCTest
@testable import GraphLinkGenerated

final class InterceptTest: XCTestCase {
    var client: GraphLinkClient!

    override func setUp() {
        client = Fixtures.newClient()
    }

    func testGreetInterceptedAppendsTheRunBeforeMarkerToTheNameArgument() async throws {
        let res = try await client.queries.greetIntercepted(name: "Ada")
        XCTAssertEqual(res.greetIntercepted.message, "Hello, Ada(runBefore)")
    }

    func testGreetPlainIsNotInterceptedNoMarkerAppended() async throws {
        let res = try await client.queries.greetPlain(name: "Ada")
        XCTAssertEqual(res.greetPlain.message, "Hello, Ada")
    }

    func testGreetingReceivedBakesTheMarkerIntoEveryEmittedGreeting() async throws {
        let stream = client.subscriptions.greetingReceived()

        let eventTask = Task {
            try await firstValue(from: stream, timeoutSeconds: 10)
        }

        try await Task.sleep(nanoseconds: 300_000_000)

        _ = try await client.mutations.sendGreeting(name: "Ada")

        let event = try await eventTask.value
        XCTAssertEqual(event.greetingReceived.message, "Hello, Ada(runBefore)")
    }

    func testTeamMembersInterceptedMappingAllowsRoleAuth() async throws {
        let res = try await client.queries.listTeams(fieldArgs: ListTeamsFieldArgs(membersRole: "auth"))
        let members = Set(res.listTeams.first!.members.map { $0.name })
        XCTAssertEqual(members, Set(["Ada", "Grace"]))
    }

    func testTeamMembersInterceptedMappingDeniesAnyOtherRole() async throws {
        do {
            _ = try await client.queries.listTeams(fieldArgs: ListTeamsFieldArgs(membersRole: "wrong"))
            XCTFail("expected throw")
        } catch is GraphLinkException {
            // expected
        }
    }
}
